src/phyltr/commands/base.py

```diff
@@ -1,28 +1,75 @@
 import fileinput
 import shlex
 import sys
+import argparse
+import re
 
 from phyltr.plumbing.sources import NewickParser
 from phyltr.plumbing.sinks import NewickFormatter
-from phyltr.utils.phyltroptparse import OptionParser
+
 
 class PhyltrCommand(object):
 
-    parser = OptionParser("Halp!")
     source = NewickParser
     sink = NewickFormatter
+    __options__ = []  # Derived classes should add parser arguments here.
+    __opt_dests = []  # When initializing an ArgumentParser, we store the argument dests here.
+
+    def __init__(self, _opts=None, **kw):
+        """
+        Command options can be initialized in two ways:
+
+        1. By passing an `argparse.Namespace` object as `_opts`.
+        2. By passing values for individual options a keyword arguments.
+        """
+        if _opts:
+            self.opts = _opts
+        else:
+            # We initialize the options with the default values of the ArgumentParser:
+            p = self.parser()
+            defaults = {d: p.get_default(d) for d in self.__opt_dests}
+            defaults.update(kw)
+            self.opts = argparse.Namespace(**defaults)
+
+    @classmethod
+    def parser(cls):
+        """
+        Assembles a pre-subcommand ArgumentParser, with the command-specific options added.
+
+        :return: ArgumentParser
+        """
+        cls.__opt_dests = []
+        prog = 'phyltr {0}'.format(cls.__name__.lower())
+        usage = """
+    {0}  [<optional arguments>] [<files>]
+
+{1}
+
+files:
+  A whitespace-separated list of filenames to read treestreams from.
+  Use a filename of "-" to read from stdin.  If no filenames are
+  specified, the treestream will be read from stdin.
+""".format(prog, re.sub('\s+', ' ', cls.__doc__.strip()))
+        res = argparse.ArgumentParser(
+            usage=usage,
+            prog='phyltr {0}'.format(cls.__name__.lower()),
+            formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+        )
+        for args, kw in cls.__options__:
+            cls.__opt_dests.append(res.add_argument(*args, **kw).dest)
+        return res
 
     @classmethod 
     def init_from_opts(cls, options, files):
-        return cls()
+        return cls(_opts=options)
 
     @classmethod 
     def run_as_script(cls):
         # Parse the arguments.
         # If there's an error, let optparse kill the process in its usual
         # fashion, as we should only be in run_as_script if we're genuinely
         # running from an interactive shell.
-        options, files = cls.parser.parse_args(exit_on_error=True)
+        options, files = cls.parser().parse_known_args()
```

Comment:
The `exit_on_error` bit is the only thing I didn't figure out yet for `argparse`.

Reply:
It's not as simple as using `ArgumentParser.parse()` instead of `ArgumentParser.parse_known_args()`?